### PR TITLE
Add option to authenticate to secure assets

### LINKED ISSUE #100: Support secure assets
### Motivation

Currently, for environments with [secure assets delivery](https://kontent.ai/learn/docs/security/secure-access/javascript#a-retrieve-assets-securely) enabled, the resulting backup contains assets with a file size of 0 kB in the `assets/` folder.

This is critical as for certain users, the backup is incomplete, and no error/warning is provided.

Preposition
Support could be added by providing the `bearer token` to the fetching of assets, I suppose:
https://github.com/kontent-ai/data-ops/blob/9da88a5618749d31e88264297da2b27f700ca1fe/src/modules/backupRestore/backupRestoreEntities/entities/assets.ts#L74

I am in general wondering if the support for secure assets would better be implemented into another package.
Also, I did not fully comprehend which places else or which modules would be affected (e.g. `migrations`?).

I will attach a PR to highlight my preposition.
The way I handled the options is not very elegant, I appreciate suggestions to improve.
In case you accept contributions on this request, I would be happy to collaborate.

There might be blind spots in the existing tests regardless of using secure assets or not, which I did not fully evaluate:
This is apparently because only `*.json` files of the backup are tested. The tests never assert the existence or size of the `assets/` folder or its contents, as far as I can tell.

### Proposed solution

Investigate the impact of adding support for secure assets, considering if this should be a core feature, if other @kontent-ai packages would benefit.
Simplest:
- Add an optional argument to provide the API key for environments where secure asset delivery (_Delivery_ API) is enabled.
- Implement error/warning at runtime if assets are not backed up correctly.
- Include tests to ensure support of secure assets.

Notes about tests, regardless of secure assets:
Assertions for asset files appear to be missing; Assert size of assets folder or better individual file size within assets folder or - even better - compare bitwise.

### Additional context

- Naming of the parameter proposed as `secureAssetDeliveryKey`
  - This is lengthy, but I think it must be the bearer for the delivery API (users might also have another token for the preview delivery API, and a shorter name could lead to confusion)
  - `Api` is not included
  - Should mentions of the key in the documentation use title case (similar to the Management API key)?
- [Retrieve assets securely](https://kontent.ai/learn/docs/security/secure-access/javascript#a-retrieve-assets-securely)

### PR DESCRIPTION
**Pull request by andgdk**
### Motivation

Currently, for environments with [secure assets delivery](https://kontent.ai/learn/docs/security/secure-access/javascript#a-retrieve-assets-securely) enabled, the resulting backup contains assets with a file size of 0 kB in the assets/ folder.

This is critical as for certain users, the backup is incomplete, and no error/warning is provided.

Preposition
Support could be added by providing the bearer token to the fetching of assets.

In addition, this PR adds the quality query parameter `?p=100` to ensure maximum quality during backup.

Only ran unit tests.

### Checklist

- [x] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [x] Docs have been updated (if applicable)
- [x] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Set up a new environment and contact Kontent.ai support to enable secure asset delivery for at least the `delivery API`.

Partially resolves #100

### CODE REVIEW

**Comment by andgdk** [comment]
Investigating a way to surface warnings to users lead me to the following investigation:
`d4d07406-adfd-404b-bbd5-58c5de6adfd2-kai-hp-highlight-3.jpg` reports a warning when comparing the blob size with the size reported by it's metadata.

`d4d07406-adfd-404b-bbd5-58c5de6adfd2-kai-hp-highlight-3.`__`jpg`__
- `backup.zip`: 27,093 bytes (_baseline_)
- `Kontent.ai` app UI: 26.45 kB (26.45 kB: 27084.8 bytes; 26.46 kb: 27095.04 bytes) (__probably rounding error__)
- Running `backupEnvironment` method:
  - CLI: 27,002 bytes (`res.size`) (__nok__)
  See PR snippet below[^1]
- `2024-11-24-10-21-backup-<UUID>.zip`: 27,002 bytes (__nok__)

I uploaded some random `.jpeg` and `.png` from my disk. Running `backupEnvironment` with the additional files reports errors on those files as well. One with significant difference: Size mismatch: expected 385_657 bytes, got 291_500 bytes (.jpeg).

Because of those results, I have to question my approach in comparing the file size. However, my expectation would still be that `res.size` should always be equal to `asset.size`. Perhaps there is a compression step which is not reflected in the metadata?
What is your take on this @IvanKiral

[^1]: https://github.com/andgdk/kontent-ai-data-ops/blob/2f880a6bf05e925b98992eb7dcae83108f638093/src/modules/backupRestore/backupRestoreEntities/entities/assets.ts#L83-L102

**Comment by IvanKiral** [comment]
Well, it seems that there is indeed a compression step. I investigated URLs, and they differ for downloading the asset and the public URL for the asset in the UI. Unfortunately, it seems that this is how a product works, and right now, backing up an asset using this tool would mean that it is being compromised (probably some metadata is missing:/). The best we can do is submit feedback on the product if you are interested. 

About secure assets - the use case is not utilized by that many people as far as I know, however, we might merge your solution with slight changes if you would like to. It would mean reverting the last commit and probably in the documentation add some information about the newly added parameter, and probably a link to the Kontent. ai learn (https://kontent.ai/learn/docs/security/secure-access/javascript#a-retrieve-assets-securely). What do you think?

**Comment by andgdk** [comment]
Hi @IvanKiral,
just pushed the requested changes, please have a look. Please be aware that I did not run integration tests for this PR.

I did not add the result of your investigation regarding a potential compression/metadata modification to the readme.
However, the known limitation section at the end of the `backupRestore` readme might be suited to do so - I leave it up to you.

For enterprise customers I feel the situation regarding the compression should be investigated, at least this needs clarification from my point of view. So, I can contact the support, and mention you as well, if you don't intervene. I would be also happy if you were in the position to keep track of this.

**Comment by IvanKiral** [comment]
Sorry for the delayed message. We were investigating a bit more and the best thing we can do now is to add a query parameter (?q=100). 

However, the image is still being comprised this way, but not that much (I believe some metadata is still missing). I also like your idea of adding it to the known limitation. Would you like to add it, or do you want me to review it as it is now and we will try to merge it? (so you have a contribution badge)

For more information about this you can refer to https://kontent.ai/learn/docs/apis/image-transformation-api#a-quality-parameter

**Comment by andgdk** [comment]
> Sorry for the delayed message. We were investigating a bit more and the best thing we can do now is to add a query parameter (?q=100).

Thanks for looking further into this. Did you also experiment with a combination of the [Lossless parameter](https://kontent.ai/learn/docs/apis/image-transformation-api#a-lossless-parameter)?

Unfortunately, I cannot request a review, so I ask this way for a review by you, @IvanKiral

File: README.md [hunk 1]
```diff
@@ -125,6 +125,15 @@ Below are the available commands:
 
 ### Examples
 
+**Backing up All Data from an Environment with Secure Asset Delivery Enabled**
```
Comment (IvanKiral) [inline README.md:128]: I don't think this one is necessary, moreover it doesn't have to be first, as this parameter will probably not be used as widely :) please remove it keep the description in the table in backupRestore readme.

File: src/commands/environment/backupRestore/README.md [hunk 2]
```diff
@@ -49,14 +49,15 @@ npx @kontent-ai/data-ops@latest environment backup --help
...
-| `--fileName`      | (Optional) The name of the output `.zip` file. Default is `backup.zip`.           |
-| `--include`       | (Optional) Specify entities to include in the backup.                             |
-| `--exclude`       | (Optional) Specify entities to exclude from the backup.                           |
-| `--configFile`    | (Optional) Path to a JSON configuration file containing parameters.               |
+| Parameter                  | Description                                                             |
+| -------------------------- | ----------------------------------------------------------------------- |
+| `--environmentId`          | The ID of the environment you want to backup.                           |
+| `--apiKey`                 | The Management API key for the environment.                             |
+| `--secureAssetDeliveryKey` | (Optional) The secure asset delivery API key for the environment.<br />Read more about enabling secure asset delivery in the [Kontent.ai documentation](https://kontent.ai/learn/docs/security/secure-access/javascript#a-retrieve-assets-securely).       |
```
Comment (IvanKiral) [inline src/commands/environment/backupRestore/README.md:56]: please move this after the --exclude parameter

File: src/commands/environment/backupRestore/README.md [hunk 3]
```diff
@@ -216,9 +228,10 @@ await restoreEnvironment(params);
 
 - **Scheduled Publishing Times**: The current API format doesn't support inclusion of the publishing time for variants scheduled to be published. The tool instead places scheduled variants into the draft step (the first step in the workflow).
 
-### Asset Size
+### Asset Limitations
 
 - **Asset Size Limit**: The Management API accepts only assets smaller than 100 MB. If your backup file contains assets larger than that (they can be uploaded through the UI), the tool won't be able to import them.
+- **Asset Quality**: Assets fetched by the tool may be compressed and metadata can be missing. You may find more information about asset compression in the [quality](https://kontent.ai/learn/docs/apis/image-transformation-api#a-quality-parameter) and [lossles](https://kontent.ai/learn/docs/apis/image-transformation-api#a-lossless-parameter) parameter documentation.
```
Comment (IvanKiral) [inline src/commands/environment/backupRestore/README.md:234]: lossles only goes for the images with webp format. Can you please remove the part mentioning it? I think that the quality parameter is enough for reference

File: src/commands/environment/backupRestore/backup.ts [hunk 4]
```diff
@@ -36,6 +36,11 @@ export const register: RegisterCommand = yargs =>
           demandOption: "You need to provide a Management API key for the given Kontent.ai environment.",
           alias: "k",
         })
+        .option("secureAssetDeliveryKey", {
+          type: "string",
+          describe:
+            "You need to provide a asset delivery key when secure asset delivery is enabled for the given Kontent.ai environment.",
```
Comment (IvanKiral) [inline src/commands/environment/backupRestore/backup.ts:42]: a secure asset delivery key

File: src/modules/backupRestore/backupRestoreEntities/entities/assets.ts [hunk 5]
```diff
@@ -69,9 +69,14 @@ const saveAsset = async (
   archive: archiver.Archiver,
   logOptions: LogOptions,
   asset: AssetContracts.IAssetModelContract,
+  secureAssetDeliveryKey: string | undefined,
 ) => {
   logInfo(logOptions, "verbose", `Exporting: file ${chalk.yellow(asset.file_name)}.`);
-  const file = await fetch(asset.url).then(res => res.blob()).then(res => res.stream());
+  const defaultOptions: RequestInit = {
```
Comment (IvanKiral) [inline src/modules/backupRestore/backupRestoreEntities/entities/assets.ts:75]: nitpick: options

File: src/modules/backupRestore/backupRestoreEntities/entityDefinition.ts [hunk 6]
```diff
@@ -11,7 +11,12 @@ export type EntityBackupDefinition<T> = Readonly<{
   displayName: string;
   fetchEntities: (client: ManagementClient) => Promise<T>;
   serializeEntities: (entities: T) => string;
-  addOtherFiles?: (loadedEntities: T, archive: archiver.Archiver, logOptions: LogOptions) => Promise<void>;
+  addOtherFiles?: (
+    loadedEntities: T,
+    archive: archiver.Archiver,
+    logOptions: LogOptions,
+    secureAssetDeliveryKey: string | undefined,
```
Comment (IvanKiral) [inline src/modules/backupRestore/backupRestoreEntities/entityDefinition.ts:18]: can you swap the logOptions and secureAssetDeliveryKey? Just for the consistency, I would like to have logOptions as last parameter :)

File: src/modules/backupRestore/backupRestoreEntities/entities/assets.ts [hunk 7]
```diff
@@ -69,9 +69,14 @@ const saveAsset = async (
...
+    headers: secureAssetDeliveryKey ? { Authorization: `Bearer ${secureAssetDeliveryKey}` } : undefined,
+  };
+  const file = await fetch(asset.url, defaultOptions)
+    .then(res => res.blob()).then(res => res.stream());
```
Comment (IvanKiral) [inline src/modules/backupRestore/backupRestoreEntities/entities/assets.ts:79]: When formatting is present, I would have all .then on new lines :)

File: src/modules/backupRestore/backupRestoreEntities/entities/assets.ts [hunk 8]
```diff
@@ -69,9 +69,14 @@ const saveAsset = async (
...
   logOptions: LogOptions,
   asset: AssetContracts.IAssetModelContract,
+  secureAssetDeliveryKey: string | undefined,
 ) => {
   logInfo(logOptions, "verbose", `Exporting: file ${chalk.yellow(asset.file_name)}.`);
-  const file = await fetch(asset.url).then(res => res.blob()).then(res => res.stream());
+  const defaultOptions: RequestInit = {
+    headers: secureAssetDeliveryKey ? { Authorization: `Bearer ${secureAssetDeliveryKey}` } : undefined,
+  };
+  const file = await fetch(asset.url, defaultOptions)
```
Comment (IvanKiral) [inline src/modules/backupRestore/backupRestoreEntities/entities/assets.ts:78]: could you please update the URL with the query parameter `?q=100`? You could use something like string formatting to add it (\`${asset.url}?q=100\`) The images would have better quality then :)

**Review by IvanKiral — CHANGES_REQUESTED** [review]

**Review by IvanKiral — APPROVED** [review]

**Comment by IvanKiral** [comment]
Thanks for the PR, I merged it will release it soon, with some other updates :)

**Comment by andgdk** [comment]
@IvanKiral Hi, just wanted to share one finding for documentation purpose.
This PR added the quality query parameter `?q=100` to the asset URL. This was requested to improve the asset quality and circumvent potential loss of quality.

Finding:
The file size can increase significantly in some instances, while the approach seam to yield good results most of the time.
Here is an example for increased file size from the `backup.zip`:
```
Exporting: file Screenshot_24-11-2024_121847_github.com.jpeg.
Size mismatch: expected 385657 bytes, got 597346 bytes.
```

**Pull request merged**